### DISCUSSION TITLE
Using a custom layout

### DISCUSSION THREAD

**Discussion by cspecter** [directus/directus#10797] [Archive]
How does one use a Custom Layout. Let's say we have defined something like this:

```javascript
import { ref } from 'vue';
import LayoutComponent from './layout.vue';

export default {
	id: 'custom-tabular-image',
	name: 'Tabular Image Layout',
	icon: 'box',
	component: LayoutComponent,
	slots: {
		options: () => null,
		sidebar: () => null,
		actions: () => null,
	},
	setup() {
		const name = ref('Custom Layout');

		return { name };
	},
};
```

How do we use that in our Interfaces? Is the name globally scoped? Do we use the ID as 'is' in a `component`?  As `<custom-tabular-image />`? Or some other way.

I see we don't have access to `useLayout` from the extensions SDK. Is there another way to access these?

The documentation is very light in this regard.

**Comment by nickrum** [comment]
The main layout component is globally registered as `layout-<id>`, so in your case you can use it as `<layout-custom-tabular-image />`.

Exposing all of the registered extensions including layouts through the extensions SDK is already planned but hasn't been implemented yet.

**Reply by cspecter** [reply]
Ok this worked. But in order to get it to show properly I had to re-create the `useLayout` composable in my interface with a stubbed out version of my new layout. This might be because I pulled in the `drawer-collection` view in order to modify it. 

In general I am finding the extension system fairly easy to use, though I am more of a React dev than Vue. The main issue I am running into now is just how to set up a more convenient development process. Right now I am pulling in `pino` into all my extensions in order to get logging to show up. But the need to build each extension and copy it over to a working Directus install really slows everything down when testing new code.

Any insight there would be much appreciated!

**Reply by nickrum** [reply]
`useLayout()` needs to have access to all of the registered layout extensions. Once we make it possible to access those from extensions, we can also expose `useLayout()` through extensions-sdk.

I fully agree that the dev workflow is quite cumbersome right now. You could use the `--watch` flag to automatically rebuild each extension if the code changed and you could set the output folder to the extensions folder (either by setting `directus:extension.path` in package.json or by using the `-o` parameter). In addition to that, you could use nodemon to restart directus whenever one of the built extension entrypoints changes.
That being said, we are actively working on an "extension dev environment" to make all of this a lot easier. This is a tricky problem to solve though, so it will likely still take some time until it's done.

**Reply by cspecter** [reply]
Got it. Another quick question. So it looks like the `useLayout` function creates a new component that appends `wrapper`, i.e. `layout-tabular-wrapper`. This component includes the setup code from the `defineLayout` function. Is there a way to get access to this `layout-tabular-wrapper` component from a layout extension in the current system? This is the one thing we have had to duplicate in our interface in order to get the layout extension to work.

**Reply by nickrum** [reply]
The wrapper component itself is generated on the fly based on the layout id passed to `useLayout()`, so you can't really access the wrapper without having access to `useLayout()`.
Normally, only the parent component that renders the layout has direct access to the wrapper component. The layout component only has access to the state defined in `setup()` in addition to the props passed to the wrapper component.

Edit: With #10849 and #10850, what you are trying to do should be a lot easier.

**Reply by michaeltoohig** [reply]
@cspecter @nickrum is there any documentation to use `useLayout` or an example from your own extension showing how to use it?
